### DISCUSSION TITLE
Is this a replacement to Intel Graphics Driver? Or an add-on on top of it?

### DISCUSSION THREAD

**Discussion by Unknown78** [pal1000/mesa-dist-win#171] [Q&A] [answered]
I don't understand quite well the essence concept of this software.

![Screenshot (6)](https://github.com/pal1000/mesa-dist-win/assets/25132599/5a6b7508-bfea-4310-bff8-f0c4ec20e559)

The CPU is i3-2310M with its integrated Intel HD Graphics 3000, and I want to unlock its full capabilities with OpenGL 3.3

The driver may be the issue that prevents me from having OpenGL 3.3
A reddit [post](https://www.reddit.com/r/lowendgaming/comments/1258usx/comment/jjzn6ri/) claim to have OpenGL 3.3 on Linux.

The graphics driver that came from Windows Update on Windows 11 only supports OpenGL 1.1
![1](https://github.com/woelper/oculante/assets/25132599/15608509-fab5-4e6f-ad92-f9a0b5e23cb5)

With an [unofficial driver](https://winraid.level1techs.com/t/get-your-intel-hd3000-sandy-bridge-windows-10-64bit-drivers-here/40044) and its [tutorial](https://www.reddit.com/r/lowendgaming/comments/1258usx/updated_win1011_drivers_for_intel_hd_3000/), I have OpenGL 3.1 now.
The developer of it claims in this [post](https://winraid.level1techs.com/t/get-your-intel-hd3000-sandy-bridge-windows-10-64bit-drivers-here/40044/31) that the hardware supports OpenGL 3.3 but Intel just don't deploy the driver of it.

![2](https://github.com/woelper/oculante/assets/25132599/cd206882-ed86-4d85-86e5-b8ad6f9470d8)

<hr>

My question:

1. Do I need to uninstall the unofficial driver again with [Wagnardsoft Display Driver Uninstaller (DDU)](https://www.wagnardsoft.com/display-driver-uninstaller-DDU-) in safe mode and replace it? Or just install this windows mesa thing on top of it?
2. How to apply it for the whole system wide? Does setting it with the command below for the system environment variable suffice?
   `setx /m MESA_GL_VERSION_OVERRIDE=3.3COMPAT`
3. What about `MESA_GLSL_VERSION_OVERRIDE`? Is it `setx /m MESA_GLSL_VERSION_OVERRIDE=330`?
4. What about `MESA_GLES_VERSION_OVERRIDE`? Which value should I set?
5. How about hardware acceleration? Does that mean will I get hardware for up to 3.1, and then just software for 3.2 and 3.3?

**Comment by pal1000** [comment]
1. mesa-dist-win is not a GPU driver replacement so you should keep it; it's an alternative and 3D graphics feature set extender .
2. Current Mesa releases (23.3.6 and 24.0.1) don't need API versions overriding anymore as they provide OpenGL and GLES support higher than your requirements on all contexts now on all drivers included.
3. Same as 2.
4. Same as 2.
5. Yes by using per app deployment tool with software requiring OpenGL >=3.2.

**Reply by Unknown78** [reply]
> 5. Yes by using per app deployment tool with software requiring OpenGL >=3.2

If I use system wide deployment, is there an automatic mechanism? 
Such that having default hardware acceleration whenever possible otherwise fallback to software rendering?

Intel HD Graphics 3000 seems to support native hardware acceleration up to OpenGL 3.3, it's just that Intel didn't deploy it.
So, for example, if an app requires <= 3.3 then choose hardware acceleration. Otherwise if it is > 3.3 then choose software rendering.

I don't want to use per app since it is tiring to set it.

**Comment by pal1000** [comment]
> If I use system wide deployment, is there an automatic mechanism?
Such that having default hardware acceleration whenever possible otherwise fallback to software rendering?

There is no automatic fallback mechanism unfortunately otherwise I wouldn't have developed a per app deployment tool.

If you could disable GPU acceleration per application then system wide deployment would have helped.

**Reply by Unknown78** [reply]
>There is no automatic fallback mechanism unfortunately otherwise I wouldn't have developed a per app deployment tool.

Would that mean system-wide deployment is going to render in software for every app from OpenGL 1.1 till the latest?

> If you could disable GPU acceleration per application then system wide deployment would have helped.

I want to use hardware acceleration whenever possible, how disabling GPU acceleration per application makes any sense?
Why it sounds counterintuitive?

**Comment by pal1000** [comment]
> Would that mean system-wide deployment is going to render in software for every app from OpenGL 1.1 till the latest?

No. System wide OpenGL deployment only has effect when GPU OpenGL driver doesn't load.  If a program needs OpenGL > 3.1 it would still fail as usual if you are using system wide OpenGL deployment.

**Reply by Unknown78** [reply]
> If a program needs OpenGL > 3.1 it would still fail as usual if you are using system wide OpenGL deployment.

Forgive me, but this is so confusing for me.

With: https://www.tablesgenerator.com/text_tables
I've made a table of it because I'm confused, 

Per-app and system-wide should be the same right? Or do we have separate table for it?
Please correct me.

```
+----+-----------------------------------+---------------+--------------------------+---------------------------+
| No | App Hardware Acceleration Setting | Native OpenGL |       Mesa OpenGL        |          Meaning          |
|    |                                   |               | (Per-app or System-wide) |                           |
+----+-----------------------------------+---------------+--------------------------+---------------------------+
| 1  | Disable                           | Fail          | Fail                     | App crash                 |
+----+-----------------------------------+---------------+--------------------------+---------------------------+
| 2  | Disable                           | Fail          | Loaded                   | Mesa software rendering   |
+----+-----------------------------------+---------------+--------------------------+---------------------------+
| 3  | Disable                           | Loaded        | Won't Load               | Native software rendering |
+----+-----------------------------------+---------------+--------------------------+---------------------------+
| 4  | Enable                            | Fail          | Fail                     | App crash                 |
+----+-----------------------------------+---------------+--------------------------+---------------------------+
| 5  | Enable                            | Fail          | Loaded                   | Mesa hardware rendering   |
+----+-----------------------------------+---------------+--------------------------+---------------------------+
| 6  | Enable                            | Loaded        | Won't Load               | Native hardware rendering |
+----+-----------------------------------+---------------+--------------------------+---------------------------+
```

**Comment by pal1000** [comment]
| GPU OpenGL driver  | Mesa OpenGL per app deployed | Mesa OpenGL system-wide deployed | OpenGL used |
---------------------- | --------------------------------- | --------------------------------------- | -------------- |
| Disabled*                  | No                                              | No                                                       | Microsoft        |                |
| Disabled*                  | No                                              | Yes                                                      | Mesa               |
| Disabled*                  | Yes                                              | No                                                      | Mesa               |
| Disabled*                  | Yes                                              | Yes                                                      | Mesa               |
| Enabled                    | No                                               | No                                                      | GPU driver       |
| Enabled                    | No                                               | Yes                                                      | GPU driver       |
| Enabled                    | Yes                                               | No                                                      | Mesa                |
| Enabled                    | Yes                                               | Yes                                                      | Mesa                |

*: Also applies to systems without GPU.

**Reply by Unknown78** [reply]
Which one that is hardware rendering? Could Mesa be also hardware rendering?

**Reply by pal1000** [reply, marked as answer]
When GPU driver is used you get hardware acceleration otherwise you don't. zink and d3d12 Mesa drivers support hardware acceleration but your GPU doesn't meet the requirements for it, Vulkan or SM>=6.2 D3D12 respectively.
